detect_os

```diff
@@ -0,0 +1,169 @@
+#!/bin/bash
+
+config="supported_os.yml"
+os_release=1 # /etc/os-release more reliable, so use it by default
+os_release_file="/etc/os-release"
+get_version=0 # Fetch OS name by default
+
+usage()
+{
+    echo -e "detect_os"
+    echo -e "\t Detects the OS that the system is currently using and prints it out"
+    echo -e "Options"
+
+    echo -e "\t -c/--config"
+    echo -e "\t\t The YAML configuration file used to help identify an OS (default is ./supported_os.yml)"
+
+    echo -e "\t --uname"
+    echo -e "\t\t Determine the current OS using 'uname -a' output"
+    echo -e "\t\t Combining this option with --os-release produces undefined behavior"
+
+    echo -e "\t --os-release"
+    echo -e "\t\t Determine the current OS using the /etc/os-release file"
+    echo -e "\t\t Combining this option with --uname produces undefined behavior"
+
+    echo -e "\t --os-version"
+    echo -e "\t\t Determine the current OS version"
+    echo -e "\t\t Works with --uname and --os-release"
+
+    echo -e "\t -h/--help/--usage"
+    echo -e "\t\t Displays this message"
+
+    exit 0
+}
+
+echo_stderr() {
+    echo ${@:1} > /dev/stderr
+}
+
+parse_os_release()
+{
+    if [ ! -f $os_release_file ];then
+        echo_stderr Could not find file $os_release_file, aborting
+        exit 1
+    fi
+
+    local os_id=`grep '^ID=' $os_release_file | cut -d'=' -f2 | sed -e "s/\"\|'//g"`
+    local version=`grep '^VERSION_ID=' $os_release_file | cut -d'=' -f2 | sed -e "s/\"\|'//g"`
+
+    local config_len=`yq '. | length' $config`
+    for i in `seq 0 $config_len`;do
+        local identifier=`yq -r ".[$i].identifier" $config`
+        
+         if [ $identifier = $os_id ]; then
+            if [ $get_version -eq 0 ];then
+                echo $identifier
+            else
+                echo $version
+            fi
+            exit 0
+        fi
+    done
+
+    echo_stderr Unknown os \"$os_id\", check $config to make sure it exists
```

Comment:
Need to exit out with an error code.

Reply:
What is the resoloution, does echo_stderr exit out? If not we want to bail here.

Reply:
look one line below, there is an `exit 1`

`echo_stderr` just echos to stderr, nothing more